### ISSUE TITLE
equalTo(String, boolean) -> Illegal Argument: ColumnType invalid

### ISSUE THREAD

**Issue by MariusBoepple** [realm/realm-java#3679]
#### Goal

We're using Realm in a library project. For synchronization, we're using a SyncAdapter. In the synchronization process, we got a simple query with a `equalTo(String, boolean)` selection to get all Persons whose flag `modified_offline` is true.
#### Expected Results

Getting every person which is marked as modified offline.
#### Actual Results

In some rare cases, it crashes and throws an `IllegalArgumentException: Illegal Argument: ColumnType invalid`. We got 4 crashes in the last week with a few hundred synchronizations a day. The code is executed in every synchronization.

Stack Trace:

```
#0. Crashed: SyncAdapterThread-34: 0 0 0x0000000000000000
       at io.realm.internal.TableQuery.nativeEqual(TableQuery.java)
       at io.realm.internal.TableQuery.equalTo(TableQuery.java:265)
       at io.realm.RealmQuery.equalTo(RealmQuery.java:407)
       at com.example.app.sync.SyncAdapter.syncPersons(PkSyncAdapter.java:164)
       at com.example.app.sync.SyncAdapter.syncComplete(PkSyncAdapter.java:147)
       at com.example.app.sync.SyncAdapter.onPerformSync(PkSyncAdapter.java:103)
       at android.content.AbstractThreadedSyncAdapter$SyncThread.run(AbstractThreadedSyncAdapter.java:272)

--

Fatal Exception: java.lang.IllegalArgumentException: Illegal Argument: ColumnType invalid.
       at io.realm.internal.TableQuery.nativeEqual(TableQuery.java)
       at io.realm.internal.TableQuery.equalTo(TableQuery.java:265)
       at io.realm.RealmQuery.equalTo(RealmQuery.java:407)
       at com.example.app.sync.SyncAdapter.syncDongles(PkSyncAdapter.java:164)
       at com.example.app.sync.SyncAdapter.syncComplete(PkSyncAdapter.java:147)
       at com.example.app.sync.SyncAdapter.onPerformSync(PkSyncAdapter.java:103)
       at android.content.AbstractThreadedSyncAdapter$SyncThread.run(AbstractThreadedSyncAdapter.java:272)

```
#### Steps & Code to Reproduce

Cannot reproduce it.. Seems to be a fairly simple operation.
#### Code Sample

SyncAdapter:

``` java
@Override
public void onPerformSync(Account account, Bundle extras, String authority,
                              ContentProviderClient provider, SyncResult syncResult) {
    Realm.init(getContext());
    Realm realm = Realm.getInstance(MyApp.getRealmConfig(getContext()));

    // Here's the crash:
    RealmResults<Person> persons = realm.where(Person.class)
             .equalTo(Person.MODIFIED_OFFLINE, true)
             .findAll();
```

MyApp:

``` java
public static RealmConfiguration getRealmConfig(Context context) {
    return new RealmConfiguration.Builder()
             .name(REALM_NAME)
             .modules(new MyModule())
             .encryptionKey(getRealmEncryptionKey(context))
             .schemaVersion(REALM_SCHEMA_VERSION)
             .deleteRealmIfMigrationNeeded()
             .build();
}
```

MyModule:

``` java
@RealmModule(library = true, allClasses = true)
public class MyModule { }
```

Person:

``` java
public class Person extends RealmObject {
   static final String MODIFIED_OFFLINE = "modified_offline";

    @PrimaryKey
    private String id;
    private boolean modified_offline;
    ...
```
#### Version of Realm and tooling

Realm version: `2.0.2`

Android Studio version: `2.2.2`

Which Android version and device: `6.0.1` (Samsung, Sony)

**Comment by beeender** [comment]
Which thread does `onPerformSync` run on? Is it in a separated process?

**Comment by MariusBoepple** [comment]
Yes, it is running in a separated process like the rest of the library (everything in `:remote`) - we're using a remote service with AIDL.

AndroidManifest:
````
<service
   android:name=".sync.SyncService"
   android:exported="true"
   android:process=":remote">
   <intent-filter>
      <action android:name="android.content.SyncAdapter" />
      </intent-filter>
         <meta-data
             android:name="android.content.SyncAdapter"
             android:resource="@xml/syncadapter" />
</service>
```

**Comment by Zhuinden** [comment]
Multi-process access of encrypted Realms is not supported as far as I know

**Comment by MariusBoepple** [comment]
We're using the same process everywhere, so it's not multi-process. And if that's the problem, we would have way more crashes I suppose.

**Comment by beeender** [comment]
Does this happen after you upgrade the apk with some schema changes? Or this is the first version of you apk?

**Comment by MariusBoepple** [comment]
It's not the first version, but we did not provide any migrations. We've added the `Parcelable` interface and may have changed some small things in the Trip class, but nothing with `start_address_name`. I also think the schema remained the same. In the Realm Configuration `.deleteRealmIfMigrationNeeded()` is set, too. The crashes occurred only after we've put everything in the remote process and made the Trip parcelable.

**Comment by beeender** [comment]
Is it getting better after more and more people update to the new version? I am asking because we have seen some issues which might be caused by the store updates -- the old apk's process is not terminated after the new apk installed and new process started.

**Comment by beeender** [comment]
And did the location/name of Realm change?

**Comment by MariusBoepple** [comment]
It's not the first version of the app but we did not change the `Person` class. Some other Realm classes have changed (e.g. added Parcelable interface) but we never provided any migration. In the configuration `.deleteRealmIfMigrationNeeded()` is set, too. But the crashes occurr only after we've put everything in the remote process.

**Comment by MariusBoepple** [comment]
The location and name did not change

**Comment by MariusBoepple** [comment]
I have got another issue which also appeared after we've put everything in a remote process: #3680

**Comment by beeender** [comment]
They probably share the same root cause... Can I get an apk to try if I can reproduce it? If it can not be accessed publicly, you can send it to `help@realm.io`. And please let me know if it may require an account to reproduce.

**Comment by MariusBoepple** [comment]
It really seems to be something with the changed process and installations. I cannot give you the APK but I will check if it's getting better and give an update then.

**Comment by Zhuinden** [comment]
Do you use Realm **anywhere** in the default process, outside of the `:remote` process?

**Comment by MariusBoepple** [comment]
Yes, every component of the app has the tag `process=":remote"`.

It may be of interest that we've also got 2 of these crashes on the day when the app update was published. Ever since this crash did not appear. This is not the library with the remote process - that's the main app which uses the default process.
````
#0. Crashed: SyncAdapterThread-1: 0 0 0x0000000000000000
       at io.realm.internal.SharedRealm.nativeGetSharedRealm(SharedRealm.java)
       at io.realm.internal.SharedRealm.getInstance(SharedRealm.java:178)
       at io.realm.internal.SharedRealm.getInstance(SharedRealm.java:155)
       at io.realm.RealmCache.createRealmOrGetFromCache(RealmCache.java:124)
       at io.realm.Realm.getDefaultInstance(Realm.java:210)
       at com.example.mainapp.MainSyncAdapter.onPerformSync(MainSyncAdapter.java:67)
       at android.content.AbstractThreadedSyncAdapter$SyncThread.run(AbstractThreadedSyncAdapter.java:272)

--

Fatal Exception: io.realm.exceptions.RealmError: Unrecoverable error. /data/data/com.example.mainapp/files/mainapp.realm: Encrypted interprocess sharing is currently unsupported.SharedGroup has been opened by pid: 29603. Current pid is 4105. in /Users/cm/Realm/realm-java/realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp line 81
       at io.realm.internal.SharedRealm.nativeGetSharedRealm(SharedRealm.java)
       at io.realm.internal.SharedRealm.getInstance(SharedRealm.java:178)
       at io.realm.internal.SharedRealm.getInstance(SharedRealm.java:155)
       at io.realm.RealmCache.createRealmOrGetFromCache(RealmCache.java:124)
       at io.realm.Realm.getDefaultInstance(Realm.java:210)
       at com.example.mainapp.MainSyncAdapter.onPerformSync(MainSyncAdapter.java:67)
       at android.content.AbstractThreadedSyncAdapter$SyncThread.run(AbstractThreadedSyncAdapter.java:272)

#0. Crashed: SyncAdapterThread-1: 0 0 0x0000000000000000
       at io.realm.internal.SharedRealm.nativeGetSharedRealm(SharedRealm.java)
       at io.realm.internal.SharedRealm.getInstance(SharedRealm.java:178)
       at io.realm.internal.SharedRealm.getInstance(SharedRealm.java:155)
       at io.realm.RealmCache.createRealmOrGetFromCache(RealmCache.java:124)
       at io.realm.Realm.getDefaultInstance(Realm.java:210)
       at com.example.mainapp.MainSyncAdapter.onPerformSync(MainSyncAdapter.java:67)
       at android.content.AbstractThreadedSyncAdapter$SyncThread.run(AbstractThreadedSyncAdapter.java:272)
```
So there really seems to be an issue with the processes and installations. Is there already an issue tracking this?

**Comment by Zhuinden** [comment]
The question is whether you do anything in `application.onCreate()`

**Comment by MariusBoepple** [comment]
I got `Realm.init(this);` there. And I discovered that the `process=":remote"` tag was missing in the `<application>` element.

**Comment by Zhuinden** [comment]
Application onCreate is called for every process that starts. This can easily break your application.

**Comment by MariusBoepple** [comment]
Calling `Realm.init()` exclusively in the remote process now. Thanks @Zhuinden. I'll give an update if it's getting better.

**Comment by kneth** [comment]
@MariusBoepple Please give us a status after your update.

**Comment by MariusBoepple** [comment]
The issue is still appearing with the fixes... I got 10 users and approx. 40 crashes (mostly there is one crash per user, only one user has more than 30 crashes). For 6 of them, the crash appeared the first time with the new version.

Some other weird issues are still there, too:
``` 
- IllegalArgumentException: Invalid native Realm type: 15 @ .equalTo()
- IllegalArgumentException: Field 'finished': type mismatch. Was BINARY, expected [BOOLEAN] @ .equalTo()
- ArrayIndexOutOfBoundsException: columnIndex > available columns @ .equalTo()
- RealmError: Unrecoverable error. Failure when converting short string to UTF-16 error_code = 1
- RealmError: Unrecoverable error. Bad transaction log in /Users/cm/Realm/realm-java/realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp line 199
```

I assume they are all related to the multiple processes used in the app...

At the end of the week, we'll release the open Beta. The app will be completely reinstalled for every user then,

**Comment by kneth** [comment]
You have to be careful as we currently don't support that. If you do a complete reinstall later this week, please tell us if your user experience the issue.

**Comment by kneth** [comment]
@MariusBoepple How did your open Beta go?

**Comment by MariusBoepple** [comment]
The problem is that we could't force the users to deinstall the old Closed-Beta app (it just shows a toast and finishes/crashes) and we assume that there are still some guys which did not deinstall it, because there are still 1-10 users affected by the issues which seems to be constant - the total number of users increased heavily (from hundred active users per day to a thousand). So for now, you can close this issue - I'll reopen it, if the issues appear again and I'm sure that our multi-process-thing works correctly 😄  by the way, it there a timeline for (encrypted) multi-process support?

**Comment by kneth** [comment]
Thanks for the update
